tests/utils/test_distributed_utils.py

```diff
@@ -0,0 +1,851 @@
+            blocksize="1MB",
+            files_per_partition=None,  # Explicitly set to None to avoid conflict
+        )
+        mock_read_blocksize.assert_called_once()
+
+        # Test reading jsonl with files_per_partition
+        result = read_data(
+            ["file1.jsonl", "file2.jsonl"],
+            file_type="jsonl",
+            files_per_partition=1,
+            blocksize=None,  # Explicitly set to None to avoid conflict
+        )
+        mock_read_fpp.assert_called_once()
+
+        # Test error handling for mixed file extensions
+        with patch("os.path.splitext", side_effect=[(".jsonl",), (".parquet",)]):
+            with pytest.raises(RuntimeError):
+                read_data(["file1.jsonl", "file2.parquet"], file_type="jsonl")
+
+        # Test invalid file type
+        with pytest.raises(RuntimeError):
+            read_data(["file.txt"], file_type="txt")
+
+
+class TestDataWritingFunctions:
+    def test_single_partition_write_with_filename(self, temp_dir):
+        """Test single_partition_write_with_filename function."""
+        # Create output directory
+        os.makedirs(os.path.join(temp_dir, "output"), exist_ok=True)
+
+        # Create test dataframe
+        df = pd.DataFrame(
+            {"text": ["hello", "world"], "file_name": ["file1.jsonl", "file2.jsonl"]}
+        )
+
+        # Test with empty dataframe
+        empty_df = pd.DataFrame(columns=df.columns)
+        result = single_partition_write_with_filename(
+            empty_df, os.path.join(temp_dir, "output"), filename_col="file_name"
+        )
+        assert result[0]
+
+        # Test with non-empty dataframe, jsonl output
+        result = single_partition_write_with_filename(
+            df,
+            os.path.join(temp_dir, "output"),
+            output_type="jsonl",
+            filename_col="file_name",
+        )
+        assert not result[0]
+        assert os.path.exists(os.path.join(temp_dir, "output", "file1.jsonl"))
+        assert os.path.exists(os.path.join(temp_dir, "output", "file2.jsonl"))
+
+        # Test with non-empty dataframe, parquet output
+        os.makedirs(os.path.join(temp_dir, "output2"), exist_ok=True)
+        result = single_partition_write_with_filename(
+            df,
+            os.path.join(temp_dir, "output2"),
+            output_type="parquet",
+            filename_col="file_name",
+        )
+        assert not result[0]
+        assert os.path.exists(os.path.join(temp_dir, "output2", "file1.parquet"))
+        assert os.path.exists(os.path.join(temp_dir, "output2", "file2.parquet"))
+
+        # Test with unknown output type
+        with pytest.raises(ValueError):
+            single_partition_write_with_filename(
+                df,
+                os.path.join(temp_dir, "output"),
+                output_type="unknown",
+                filename_col="file_name",
+            )
+
+    @patch("nemo_curator.utils.distributed_utils._write_to_jsonl_or_parquet")
```

Comment:
I think there may be blocks of `_write_to_jsonl_or_parquet` that are not covered?